### DISCUSSION TITLE
关于duilib显示一个模态窗口后程序仍往下执行的问题

### DISCUSSION THREAD

**Discussion by QHHNB** [netease-im/NIM_Duilib_Framework#421] [General]
之前开发一直都是用的原生的duilib，最近要换到网易的duilib，有这样一个功能：
先弹出一个提示框，然后根据提示框的选择再进行下一步操作，之前是使用一个模态窗口来实现。
原理就是模态窗口show了之后程序会停止然后等待模态窗口的操作再继续运行，于是我在试验网易duilib 的msgbox控件的时候，我把它原本的showwindow()函数改成了showmodalfake(hwnd);
![image](https://user-images.githubusercontent.com/106065825/172005553-bea620e1-8e13-4ecf-9adf-66cc972796e5.png)
按照设想，我在显示弹出提示框的下一步加上断点
![image](https://user-images.githubusercontent.com/106065825/172005590-3d4da8ed-9da3-42e1-a63a-7a407c14bc69.png)
然后点击listitem，它应该等我关闭这个模态窗口后才会执行到return这一步。但是实际上点击后，窗口还没弹出，就已经执行到return这一步。
![image](https://user-images.githubusercontent.com/106065825/172005931-07880b64-ac22-4b24-9754-86fccf86185e.png)
于是我进入showmodalfake(hwnd);发现它的实现如下：
![image](https://user-images.githubusercontent.com/106065825/172006268-72fcf18e-9c12-4d70-a8eb-adde259adb10.png)
对比之前我使用的原生duilib的showmodal()
![image](https://user-images.githubusercontent.com/106065825/172006325-1080dd56-742f-4f0f-8b77-ab240049e3ad.png)
两者一对比，发现是少了一个while的过程，于是对网易的showmodalfake(hwnd)做了以下修改
![image](https://user-images.githubusercontent.com/106065825/172006444-dba7f826-edb4-47a4-9eee-b574784ef937.png)
解决了这个问题，实现了只有模态窗口操作后后才能进行下一步操作。
但是不知道这样是否合理，望指正。

**Comment by nmgwddj** [comment]
不推荐这样使用，while 会导致整个主线程被阻塞，除了 msgbox 显示以外，其他窗口的渲染应该正常执行而不是被一个 msgbox 阻塞掉。这个 msgbox 是一个异步的，当点击确定或者取消后，投递新的任务到 UI 线程来执行任务，而不应该阻塞整个主线程。

**Comment by QHHNB** [comment]
非常感谢




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "netease-im/NIM_Duilib_Framework"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月6日(星期一) 上午9:37
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [netease-im/NIM_Duilib_Framework] 关于duilib显示一个模态窗口后程序仍往下执行的问题 (Discussion #421)





 
不推荐这样使用，while 会导致整个主线程被阻塞，除了 msgbox 显示以外，其他窗口的渲染应该正常执行而不是被一个 msgbox 阻塞掉。这个 msgbox 是一个异步的，当点击确定或者取消后，投递新的任务到 UI 线程来执行任务，而不应该阻塞整个主线程。
